projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-admin-menu.php

```diff
@@ -404,6 +416,105 @@ private function parse_menu_item( $title ) {
 
 		return $item;
 	}
+
+	/**
+	 * Check if the given string is structured like a domain
+	 *
+	 * @TODO: Support deep nested domains like jp.eu.foo.something.something.wordpress.com
+	 * @param string $domain_name_value The string that will be checked.
+	 *
+	 * @return false|int
+	 */
+	private function is_valid_domain_name( $domain_name_value = '' ) {
+		return preg_match( '/^(http[s]?\:\/\/)?((\w+)\.)?(([\w-]+)?)(\.[\w-]+){1,2}$/', $domain_name_value );
+	}
+
+	/**
+	 * Generate a identifier based on menu or submenu slug
+	 *
+	 * @param string $key Menu or submenu slug.
+	 *
+	 * @return string
+	 */
+	private function generate_identifier( $key ) {
+		$replace_table = array(
+			'http://'        => '',
+			'https://'       => '',
+			'.php'           => '', // we add this in order to avoid cases where customizer.php can be considered a domain.
+			'toplevel_page_' => '',
+		);
+
+		// Replace strings with the ones from the $replace_table.
+		$path_with_table_strings = strtr( $key, $replace_table );
+
+		// Split string on '/' or '?' to account for query strings.
+		$array_of_paths = preg_split( '/(\/|\?)/', $path_with_table_strings, -1, PREG_SPLIT_NO_EMPTY );
+
+		// Dont touch domain names as these get filtered out next and replace non-alphanumeric characters with an underscore.
+		$parsed_query_strings = $this->format_non_domain_identifiers( $array_of_paths );
+
+		$array_without_domains = array_filter(
+			$parsed_query_strings,
+			function ( $value ) {
+				return ! $this->is_valid_domain_name( $value );
+			}
+		);
+
+		return sanitize_key( implode( '_', $array_without_domains ) );
+	}
+
+	/**
+	 * Format non-domain identifiers
+	 *
+	 * It replaces non alphanumeric characters with underscores in paths that don't follow a domain-like structure.
+	 *
+	 * e.g. Given an array like ['test-test, 'example-domain.com'] will become ['test_test', 'example-domain.com']
+	 *
+	 * @param array $paths An array of slug parts.
+	 *
+	 * @return array|null[]|string[]|string[][]
+	 */
+	private function format_non_domain_identifiers( $paths ) {
+		return array_map(
+			function ( $value ) {
+				if ( $this->is_valid_domain_name( $value ) ) {
+					return $value;
+				}
+
+				return preg_replace( '/[^a-z0-9]+/', '_', strtolower( urldecode( $value ) ) );
+			},
+			$paths
+		);
+	}
+
+	/**
+	 * Map known custom identifiers to the newer version
```

Comment:
Nitpick: Known where or by who? This seems to transform a list of menu identifiers to a list of event names which we want to use in Calypso. Can you rename the function/add extra comments to better denote that? 

Example:

```php
/**
 * Returns an array that maps menu identifiers to Calypso events names.
 */ 
private function get_calypso_events_mapping() {

}
```

Reply:
Hmm... good point, but we don't return the entire event name we use in Calypso, just the unique part of it. I'll change the name to something like `get_calypso_event_id_mapping`.